File: lib/EqCommands.ml
open FOL
open LK
open Claire
open Env
open Lexer

exception WrongArgument of argument

let refl : env -> argument -> judgement list -> command list =
  fun env arg js -> match arg with
  | ArgTerms [t] -> 
    [ Apply [Cut(pFormula "Forall r. eq(r,r)")];
      Use("refl", ["eq", PredFun(["x"], PredFml(pFormula "eq(x,x)"))]);
      Apply [ForallR "r"];
      NewCommand("assumption", ArgEmpty);
      Apply [ForallL t];
      NewCommand("assumption", ArgEmpty);
    ]
  | arg -> raise (WrongArgument arg)

let export_command = [ "refl", refl ]
let export_decl    = []

File: lib/eqCommands.pl
:- module('lib/eqCommands',[]).

refl(_,t(T),_,[
  apply([cut(forall(r,eq*[r,r]))]),
  use(refl, [eq: ([x]=>eq*[x,x])]),
  apply([forallR(r)]),
  assumption*[],
  apply([forallL(T)]),
  assumption*[]
]).
refl(_,Arg,_,_) :- throw(wrongArgument(Arg)).

export_command([refl]).
export_decl([]).
